search/materialize.go

```diff
@@ -403,63 +403,115 @@ func (m *Materializer) materializeAllLabels(ctx context.Context, rgi int, rr []R
 		attribute.Int("row_ranges_count", len(rr)),
 	)
 
-	labelsRg := m.b.LabelsFile().RowGroups()[rgi]
-	cc := labelsRg.ColumnChunks()[m.colIdx]
-	colsIdxs, err := m.materializeColumn(ctx, m.b.LabelsFile(), rgi, cc, rr, false)
+	// Get column indexes for all rows in the specified ranges
+	columnIndexes, err := m.getColumnIndexes(ctx, rgi, rr)
 	if err != nil {
-		return nil, errors.Wrap(err, "materializer failed to materialize columns")
+		return nil, errors.Wrap(err, "failed to get column indexes")
 	}
 
-	colsMap := make(map[int]*[]parquet.Value, 10)
-	results := make([][]labels.Label, len(colsIdxs))
-
-	for _, colsIdx := range colsIdxs {
-		idxs, err := schema.DecodeUintSlice(colsIdx.ByteArray())
-		if err != nil {
-			return nil, errors.Wrap(err, "materializer failed to decode column index")
-		}
-		for _, idx := range idxs {
-			colsMap[idx] = &[]parquet.Value{}
-		}
+	// Build mapping of which columns are needed for which row ranges
+	columnToRowRanges, rowRangeToStartIndex, err := m.buildColumnMappings(rr, columnIndexes)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to build collum mapping")
 	}
 
-	errGroup, ctx := errgroup.WithContext(ctx)
-	errGroup.SetLimit(m.concurrency)
+	// Materialize label values for each column concurrently
+	results := make([][]labels.Label, len(columnIndexes))
+	mtx := sync.Mutex{}
+	errGroup := &errgroup.Group{}
+	labelsRowGroup := m.b.LabelsFile().RowGroups()[rgi]
 
-	for cIdx, v := range colsMap {
+	for columnIndex, rowRanges := range columnToRowRanges {
 		errGroup.Go(func() error {
-			cc := labelsRg.ColumnChunks()[cIdx]
-			values, err := m.materializeColumn(ctx, m.b.LabelsFile(), rgi, cc, rr, false)
+			// Extract label name from column schema
+			columnChunk := labelsRowGroup.ColumnChunks()[columnIndex]
+			labelName, ok := schema.ExtractLabelFromColumn(m.b.LabelsFile().Schema().Columns()[columnIndex][0])
+			if !ok {
+				return fmt.Errorf("column %d not found in schema", columnIndex)
+			}
+
+			// Materialize the actual label values for this column
+			labelValues, err := m.materializeColumn(ctx, m.b.LabelsFile(), rgi, columnChunk, rowRanges, false)
 			if err != nil {
-				return errors.Wrap(err, "failed to materialize labels values")
+				return errors.Wrap(err, "failed to materialize label values")
 			}
-			*v = values
+
+			// Assign label values to the appropriate result positions
+			mtx.Lock()
+			defer mtx.Unlock()
+
+			valueIndex := 0
+			for _, rowRange := range rowRanges {
+				startIndex := rowRangeToStartIndex[rowRange]
+
+				for rowInRange := 0; rowInRange < int(rowRange.Count); rowInRange++ {
+					if !labelValues[valueIndex].IsNull() {
+						results[startIndex+rowInRange] = append(results[startIndex+rowInRange], labels.Label{
+							Name:  labelName,
+							Value: util.YoloString(labelValues[valueIndex].ByteArray()),
```

Comment:
We are adding labels to the final results in multiple goroutine. For each series, do we need to ensure labels are sorted or we already do it in another place?